File: Format/COM.java
package Format;

import swingIO.Descriptor;
import swingIO.tree.*;
import javax.swing.tree.*;

//Processor cores.

import core.x86.*; //X86.

public class COM extends Window.Window implements JDEventListener
{
  private static Descriptor blank = new Descriptor( file );

  public COM()
  {
    tree.setEventListener( this );

    //Root node is now the target file.
 
    ((DefaultTreeModel)tree.getModel()).setRoot(null); tree.setRootVisible(true); tree.setShowsRootHandles(true);
    
    JDNode root = new JDNode( fc.getFileName() );

    //header data folder to organize exe setup information.
  
    JDNode t = new JDNode("Info.h"); root.add( t ); root.add( new JDNode("Program Start (Machine code).h") );

    if( core == null || core.type() != 0 ){ core = new X86( file ); } else { core.setTarget( file ); }

    try { file.addV( 0, file.length(), 0x0100, file.length() ); } catch( Exception e ) { }

    core.setBit(X86.x86_16); core.setEvent( this::Dis );

    ((DefaultTreeModel)tree.getModel()).setRoot(root);

    tree.setSelectionPath( new TreePath( t.getPath() ) ); open( new JDEvent(this, "Info.h") );
  }

  public void open( JDEvent e )
  {
    if( e.getPath().equals("Info.h") )
    {
      ds.setDescriptor( blank );
      info("<html>DOS COM Files have no header or setup information. The program begins at the start of the file.</html>");
    }
    else
    {
      core.clear(); core.Crawl.add( 0x0100L ); core.disLoc(0, true); ds.setDescriptor( core );
    }
  }

  //Disassemble routine.

  public void Dis( int loc, boolean crawl )
  {
    try
    {
      //Disassemble till end, or return from application.
      //Note that more can be added here such as the jump operation.

      file.seekV( loc );

      String t1 = "", t2 = "", t = "";

      int Dos_exit = 0;

      while( true )
      {
        t1 = core.posV(); t2 = core.disASM();
        
        if( Dos_exit == 0 && ( t2.startsWith("MOV AX,4C") || t2.startsWith("MOV AH,4C") ) ) { Dos_exit = 1; }
        else if( Dos_exit == 1 && ( t2.indexOf("AX,") > 0 || t2.indexOf("AH,") > 0 ) ) { Dos_exit = 0; }
        if( Dos_exit == 1 && t2.equals("INT 21") ) { Dos_exit = 2; }
        
        t += t1 + " " + t2 + "<br />";

        if( Dos_exit == 2 || t2.startsWith("RET") || t2.startsWith("JMP") || t2.equals("INT 20") ) { break; }
      }

      info( "<html>" + t + "</html>" ); core.clean(loc, file.getVirtualPointer());

      long pos = file.getFilePointer() - 1, posV = file.getVirtualPointer() - 1;
      
      file.seekV( loc ); Virtual.setSelectedEnd( posV ); Offset.setSelectedEnd( pos );

      ds.setDescriptor( core );
    }
    catch( Exception e ) { e.printStackTrace(); }
  }
}

File: Format/com.js
//Everything in the format reader must be defined as an object called format.
//This ensures that all data in the format reader is overwritten when loading in a new format.

format = {
  //DOS com files have a blank header.

  header: new Descriptor([]),

  //Function load is always called first.

  load: function()
  {
    //DOS com Files have no header.

    this.header.setEvent(this, "dataEvent");

    //Dos com files are by default x86-16 bit programs.

    loadCore("core/x86/dis-x86.js");

    //Dos com files are loaded at 0x100 by default in virtual address space.

    file.addV( 0, file.size, 0x0100, file.size );

    //The root node is the binary application.
    
    var root = new treeNode(file.name.substring(file.name.lastIndexOf("/")+1,file.name.length),[],true);

    root.add("Info.h", [0]); root.add("Program Start (Machine code).h", [1]);
    Tree.set(root); tree.prototype.event = this.treeEvent;

    //Show virtual address space.

    if( !virtual.visible ) { showH(true); } file.seekV(0x100); virtual.sc();

    //Set the default selected node.

    tree.prototype.treeClick( Tree.getNode(0).getNode(0).parentElement );
  },

  //Tree event handling.

  treeEvent: function(node)
  {
    var args = node.getArgs();

    if( args[0] == 0 ) { dModel.setDescriptor(format.header); }
    else
    {
      core.addressMap = true; core.resetMap(); core.bitMode = 0; file.seekV(0x100);
      
      core.setCodeSeg((Math.random()*0x2000)<<3); dModel.setCore(core); dModel.coreDisLoc(0x100,true);
    }
  },

  //When user clicks on information for the program header.

  dataEvent: function()
  {
    info.innerHTML = "DOS COM Files have no header or setup information. The program begins at the start of the file and is placed at 0x100 in RAM memory.";
  }
}

//The data descriptor calls this function when we go to click on an address we wish to disassemble.

dModel.coreDisLoc = function(virtual,crawl)
{
  //We still have to move to a section and read the data if it is not loaded.
  //The function dis is called when the data is ready, or is already in position.

  if( this.dis == null ) { this.dis = function()
  {
    //Set binary code relative position within the buffer.

    core.setBinCode(file.dataV,this.vr);
  
    //Begin disassembling the code.
  
    info.innerHTML = "<pre>" + core.disassemble(this.cr) + "</pre>";
    
    file.seekV(file.dataV.offset+this.vr); dModel.adjSize(); dModel.update();
  }}

  //Begin data check.

  this.vr = virtual; this.cr = crawl; core.setAddress(this.vr);

  //If the address we wish to disassemble is within the current memory buffer then we do not have to read any data.

  var pos = this.vr - file.dataV.offset; if(pos >= 0 && pos <= file.dataV.length) { this.vr = pos; this.dis(); }
 
  //Else we need to locate to the section before disassembly can happen.
 
  else { this.vr = virtual & 0x0F; file.call( this, "dis" ); file.seekV(virtual-this.vr); file.readV(window.innerHeight >> 1); }
}